backend/server-connection/src/main/java/org/sonarsource/sonarlint/core/serverconnection/LocalStorageSynchronizer.java

```diff
@@ -74,12 +75,28 @@ public boolean synchronize(ServerApi serverApi) {
     return pluginsSynchronizer.synchronize(serverApi, new ProgressMonitor(null));
   }
 
-  public void synchronize(ServerApi serverApi, String projectKey) {
-    var analyzerConfiguration = synchronizeAnalyzerConfig(serverApi, projectKey, new ProgressMonitor(null));
-    storage.project(projectKey).analyzerConfiguration().store(analyzerConfiguration);
+  public AnalyzerConfigUpdateSummary synchronize(ServerApi serverApi, String projectKey) {
+    var updatedAnalyzerConfiguration = synchronizeAnalyzerConfig(serverApi, projectKey, new ProgressMonitor(null));
+    AnalyzerConfigUpdateSummary configUpdateSummary;
+    try {
+      var originalAnalyzerConfiguration = storage.project(projectKey).analyzerConfiguration().read();
+      configUpdateSummary = diffAnalyzerConfiguration(originalAnalyzerConfiguration, updatedAnalyzerConfiguration);
+    } catch (StorageException e) {
+      configUpdateSummary = null;
+    }
+
+    storage.project(projectKey).analyzerConfiguration().store(updatedAnalyzerConfiguration);
     var version = storage.serverInfo().read().orElseThrow().getVersion();
     serverApi.newCodeApi().getNewCodeDefinition(projectKey, null, version)
       .ifPresent(ncd -> storage.project(projectKey).newCodeDefinition().store(ncd));
+    return configUpdateSummary;
+  }
+
+  private static AnalyzerConfigUpdateSummary diffAnalyzerConfiguration(AnalyzerConfiguration original, AnalyzerConfiguration updated) {
+    var originalFileExclusions = original.getSettings().getAll().get("sonar.exclusions");
```

Comment:
I don't like to have the responsibility of knowing what are the properties involved in the exclusion process here. This should be in the `FileExclusionService`.
I would raise an event as soon as some property changed in the local storage (with a diff, or simply the old and new maps). Then the `FileExclusionService` would receive this event, and look if one of the exclusions/inclusions properties have been updated.

Reply:
Good idea. In [157701d](https://github.com/SonarSource/sonarlint-core/pull/827/commits/157701d423cb61219539615451f78474538d5cb5) we make diff on the ServerSettings and send the summary to the `SynchronizationService` which then raises `SonarServerSettingsChangedEvent`.
The reason for raising the event in `SynchronizationService` is that the `LocalStorageSynchronizer` is not a bean (and is used from `ServerConnection` which is used from the `ConnectedEngine` so it wouldn't be straightforward making it a bean)